src/slot.rs

```diff
@@ -642,102 +648,238 @@ impl Stats {
+// Provides direct access to all underlying slots
+//
+// Shared by both a [`SetWorker`] and [`Set`]
+struct Slots<Conn: Connection> {
     pool_name: pool::Name,
     name: backend::Name,
     backend: Backend,
-    config: SetConfig,
-
-    wanted_count: usize,
-
     // Interface for actually connecting to backends
     backend_connector: SharedConnector<Conn>,
+    config: SetConfig,
 
-    // Interface for receiving client requests
-    rx: mpsc::Receiver<SetRequest<Conn>>,
+    // The actual underlying slots, by ID.
+    slots: BTreeMap<SlotId, Slot<Conn>>,
+    // The desired number of slots
+    wanted_count: usize,
+    next_slot_id: SlotId,
 
-    // Identifies that the set worker should terminate immediately
-    terminate_rx: tokio::sync::oneshot::Receiver<()>,
+    // If "true", new requests are rejected
+    terminating: bool,
 
     // Interface for communicating backend status
     status_tx: watch::Sender<SetState>,
-
-    // Sender and receiver for returning old handles.
-    //
-    // This is to guarantee a size, and to vend out permits to claim::Handles so they can be sure
-    // that their connections can return to the set without error.
+    // Sender for returning old handles.
     slot_tx: mpsc::Sender<BorrowedConnection<Conn>>,
-    slot_rx: mpsc::Receiver<BorrowedConnection<Conn>>,
-
-    // The actual slots themselves.
-    slots: BTreeMap<SlotId, Slot<Conn>>,
 
     // Summary information about the health of all slots.
     //
     // Must be kept in lockstep with "Self::slots"
     stats: Arc<Mutex<Stats>>,
-
     failure_window: Arc<WindowedCounter>,
-
-    next_slot_id: SlotId,
 }
 
-impl<Conn: Connection> SetWorker<Conn> {
-    #[allow(clippy::too_many_arguments)]
-    fn new(
-        pool_name: pool::Name,
-        set_id: u16,
-        name: backend::Name,
-        rx: mpsc::Receiver<SetRequest<Conn>>,
-        terminate_rx: tokio::sync::oneshot::Receiver<()>,
-        status_tx: watch::Sender<SetState>,
-        config: SetConfig,
-        wanted_count: usize,
-        backend: Backend,
-        backend_connector: SharedConnector<Conn>,
-        stats: Arc<Mutex<Stats>>,
-        failure_window: Arc<WindowedCounter>,
-    ) -> Self {
-        let (slot_tx, slot_rx) = mpsc::channel(config.max_count);
-        let mut set = Self {
-            pool_name,
-            name,
-            backend,
-            config,
-            wanted_count,
-            backend_connector,
-            stats,
-            failure_window,
-            rx,
-            terminate_rx,
-            status_tx,
-            slot_tx,
-            slot_rx,
-            slots: BTreeMap::new(),
-            next_slot_id: SlotId::first(set_id),
+impl<Conn: Connection> Slots<Conn> {
+    #[instrument(
+        level = "trace",
+        skip(self),
+        err,
+        name = "Slots::claim",
+        fields(name = ?self.name),
+    )]
+    fn claim(&mut self, id: ClaimId) -> Result<claim::Handle<Conn>, Error> {
```

Comment:
This function has basically been moved from the `SetWorker` to this new struct, `Slots`

Reply:
The `SlotSet` used to have an `async fn claim` method that used a oneshot to communicate with the `SlotWorker`. The worker then called a synchronous `claim` function and returned the result.

This new structure avoids the message passing to the `SlotWorker`, by pulling out this `Slots` structure. Now, there's no message passing -- as you said, yeah, we're just locking `Slots` and pulling the connection out of it.